runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java

```diff
@@ -0,0 +1,249 @@
+        SMRGarbageEntry garbageEntry;
+        while ((garbageEntry = garbageReceivingQueue.poll()) != null && drainCount < BATCH_SIZE) {
+            long globalAddress = garbageEntry.getGlobalAddress();
+
+            if (!addressToGarbage.containsKey(globalAddress)) {
+                addressToGarbage.put(globalAddress, garbageEntry);
+            } else {
+                addressToGarbage.get(globalAddress).merge(garbageEntry);
+            }
+            ++drainCount;
+        }
+
+        if (drainCount == 0) {
+            log.trace("Garbage Informer has nothing to send");
+        }
+
+        GarbageBatch garbageBatch = new GarbageBatch(addressToGarbage.values(), lastMarkers, rt);
+        lastMarkers.clear();
+        garbageSendingDeque.offer(garbageBatch);
+
+        sendGarbage();
+    }
+
+    private void sendGarbage() {
+        GarbageBatch garbageBatch = null;
+        try {
+            while ((garbageBatch = garbageSendingDeque.poll()) != null) {
+                garbageBatch.sendToLogUnits();
+            }
+        } catch (Exception e) {
+            log.error("Caught exception in the write processor.", e);
+            // Adds pending garbage to the head of garbageSendingDeque and waits for another
+            // cycle to send garbage to LogUnit servers.
+            if (garbageBatch != null) {
+                garbageSendingDeque.addFirst(garbageBatch);
+            }
+        }
+    }
+
+    /**
+     * Contains a batch of garbageEntries as well as the last marker address.
+     * The batch of garbageEntries and the marker address are sent to LogUnit
+     * servers atomically, i.e. if any of them is failed to send to the LogUnits,
+     * all of them are sent again.
+     */
+    @VisibleForTesting
+    @AllArgsConstructor
+    public static class GarbageBatch {
+        private final Collection<SMRGarbageEntry> garbageEntries;
+        private final Map<UUID, Long> lastMarkers;
+        private final CorfuRuntime rt;
+
+        /**
+         * Sends GarbageEntry batch as well as last marker address to LogUnits.
+         */
+        public void sendToLogUnits() {
+            rt.getAddressSpaceView().layoutHelper(e -> {
+                Layout layout = e.getLayout();
+                Map<Layout.LayoutStripe, List<SMRGarbageEntry>> stripToGarbageEntries = new HashMap<>();
+
+                // shard SMRGarbageEntries based on stripe.
+                garbageEntries.forEach(garbageEntry -> {
+                    long globalAddress = garbageEntry.getGlobalAddress();
+                    Layout.LayoutStripe strip = layout.getStripe(globalAddress);
+
+                    stripToGarbageEntries.computeIfAbsent(strip, s -> new ArrayList<>()).add(garbageEntry);
+
+                });
+
+                // send GarbageEntry batch.
+                stripToGarbageEntries.forEach((strip, garbageEntries) ->
+                        rt.getAddressSpaceView().sparseTrim(rt.getLayoutView().getRuntimeLayout(), strip,
```

Comment:
Can this be parallelized by stripe?
If any stripe succeeds, you can remove all those SMRGarbageEntry from the `garbageEntries` in this batch.

Reply:
Do you mean using parallel stream to parallel?

Reply:
I mean just send out all the requests at once and get the futures.

Reply:
It is not easy to do the refactor. I will do it in the future.